src/Tasks/ResolveAssemblyReferences/Client/RarClient.cs

```diff
@@ -0,0 +1,67 @@
+﻿using Microsoft.Build.Framework;
+using Microsoft.Build.Shared;
+using Microsoft.Build.Tasks.ResolveAssemblyReferences.Contract;
+using Microsoft.VisualStudio.Threading;
+using StreamJsonRpc;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Pipes;
+using System.Linq;
+using System.Security.Principal;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Build.Tasks.ResolveAssemblyReferences.Client
+{
+    internal sealed class RarClient : IDisposable
+    {
+        private const int ConnectionTimeout = 30;
+        private readonly IRarBuildEngine _rarBuildEngine;
+        private NamedPipeClientStream _clientStream;
+
+        public RarClient(IRarBuildEngine rarBuildEngine)
+        {
+            _rarBuildEngine = rarBuildEngine;
+        }
+
+
+        internal bool Connect()
+        {
+            var pipeName = _rarBuildEngine.GetRarPipeName();
+            var stream = _rarBuildEngine.GetRarClientStream(pipeName, ConnectionTimeout);
+
+            if (stream == null)
+                return false; // We couldn't connect
+
+            _clientStream = stream;
+            return true;
+        }
+
+        internal bool CreateNode()
+        {
+            return _rarBuildEngine.CreateRarNode();
+        }
+
+        internal int GetNumber(int parameter)
+        {
+            using var client = GetRpcClient();
+            // TODO: Find out if there is any possibility of awaiting it.
+            return client.GetNumber(parameter).GetAwaiter().GetResult();
```

Comment:
if the object supports GetAwaiter, then it should be possible to  await it (calling it async)

Reply:
Yes, but the Execute method (from witch is this code called) is synchronous and I can't change that. I can move this behavior to that method, but it will be still somewhere.